sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -753,6 +753,9 @@ object LimitPushDown extends Rule[LogicalPlan] {
     // Merge offset value and limit value into LocalLimit and pushes down LocalLimit through Offset.
     case LocalLimit(le, Offset(oe, grandChild)) =>
       Offset(oe, LocalLimit(Add(le, oe), grandChild))
+    // Push down limit 1 if join type is LeftSemiOrAnti and join condition is empty
+    case j @ Join(_, right, LeftSemiOrAnti(_), None, _) if !right.maxRows.exists(_ <= 1) =>
```

Comment:
what's the reason to move it from `pushLocalLimitThroughJoin` to here?

Reply:
To support queries do not contain limit. For example:
```sql
SELECT * FROM v1 WHERE type IN (SELECT type FROM t2)
```

Reply:
`pushLocalLimitThroughJoin` only supports the following case:
```sql
SELECT * FROM v1 WHERE type IN (SELECT type FROM t2) LIMIT 10;
```

Reply:
we should write this down explicitly in the PR description...